RHDEVDOCS 6385 procedure for PAC pipeline runs (#89926)
openshift/openshift-docs: modules/op-creating-pipeline-run-pac.adoc

// This module is included in the following assemblies:
// * pac/creating-pipeline-runs-pac.adoc

:_mod-docs-content-type: PROCEDURE
[id="creating-pipeline-run-pac_{context}"]
= Creating a pipeline run in {pac}

You can create a pipeline run definition for {pac} in your repository and match it to an event, such as a pull request. When the event happens, {pac} creates a `PipelineRun` custom resource (CR) from this definition, and then {pipelines-shortname} executes the pipeline run.

.Prerequisites

* You configured {pac} to integrate with your Git repository hosting service provider.

* You created a `Repository` custom resource (CR) to define the connection to your repository in {pac}.

* You have a `.tekton` directory in the root of your repository.

.Procedure

. Create a file with a `.yaml` or `.yml` extension in the `.tekton` directory.

. In the file that you created, create a YAML specification for a `PipelineRun` CR. This specification can use all features of {pipelines-shortname}.

> . Depending on the requirements of your pipeline run definition, complete any number of the following optional steps:

Reviewer: ```suggestion
. Depending on the requirements of your pipeline run definition, complete any of the following optional steps:
```

+
For more information about dynamic variables, see the "Dynamic variables in a pipeline run specification" section.

** Add one or more {pac} resolver annotations to your pipeline run definition. A {pac} resolver annotation references a pipeline or task in {tekton-hub}, an HTTP location, or a location in your repository outside the `.tekton` directory. If you create a {pac} resolver annotation to reference a resource, you can use this resource by name in your pipeline run definition. The {pac} resolver automatically resolves these resources and includes them in the  `PipelineRun` CR that is based on your definition.
+
For more information about these annotations, see the "{pac} resolver annotations" section.

. Match the pipeline run to events by adding any of the following annotations to the pipeline run definition. When the defined event happens, {pac} starts the pipeline run.  For more information about matching the pipeline run to events, see the "Annotations for matching events to a pipeline run" section.

> ** A combination of the `pipelinesascode.tekton.dev/on-event` annotation, which defines a `pull request` or `push` event, and a `pipelinesascode.tekton.dev/on-target-branch` annotation, which defines the branch that the pull request or push event must target. If you match your pipeline run to a pull request or push event, the pipeline run starts when the event is created and also starts again each time the source branch of the pull request or push event is updated.

Reviewer 2: ```suggestion
** A combination of the `pipelinesascode.tekton.dev/on-event` annotation, which defines a `pull request` or `push` event, and a `pipelinesascode.tekton.dev/on-target-branch` annotation, which defines the branch that the pull request or push event must target. If you match your pipeline run to a pull request or push event, the pipeline run starts when the event is created and also triggers again each time the source branch of the pull request is updated.
```

@mramendi there is no need to include push on update because push is a single event (I meant push event doesn't get updated)

Author: I don't want to use "triggers" as it might create confusion with the component named Triggers but I am taking the other points on board, thanks

// This module is included in the following assemblies:
// * pac/creating-pipeline-runs-pac.adoc

:_mod-docs-content-type: PROCEDURE
[id="creating-pipeline-run-pac_{context}"]
= Creating a pipeline run in {pac}

Complete the following general steps to create a pipeline run definition for {pac} in your repository. Refer to subsequent sections for detailed reference information.

.Prerequisites

> * You configured {pac} to integrate with your Git repository  hosting service provider.

Reviewer 2: ```suggestion
* You configured {pac} to integrate with your Git repository hosting service provider.
```

* You created a `Repository` custom resource (CR) to define the connection to your repository in {pac}.

.Procedure

. If a `.tekton` directory does not exist in the root of your repository, create this directory.

> . Create a file with a `.yaml` or `.yml` extension in the `.tekton` directory. In this file, create a YAML specification for a `PipelineRun` CR. This specification can use all features of {pipelines-shortname}.

Author: The question of using examples here is rather interesting and I'd like to discuss it more. Building from a basic excampel and adding parts one by one might be interesting but (a) might make the text bulky/jhard to read and (b) at some points the examples will not be valid.

> . Create a file with a `.yaml` or `.yml` extension in the `.tekton` directory. In this file, create a YAML specification for a `PipelineRun` CR. This specification can use all features of {pipelines-shortname}.

Reviewer: Yeah, there are multiple approaches - happy to chat

> . Optional: Use parameters in your pipeline run specification. The {pac} resolver replaces these parameters with values that represent the current context. For example, `{{ repo_url }}` is the current URL for the repository and `{{ revision }}` is the branch on which the pipeline run was started. See the "Parameters in a pipeline run specification" section for a full list of the parameters.

Reviewer 3: > See the "Parameters in a pipeline run specification" section for a full list of the parameters.

When you say  `See the Parameters in a pipeline run specification` how about giving a link to redirect :thinking:

> . Optional: Use parameters in your pipeline run specification. The {pac} resolver replaces these parameters with values that represent the current context. For example, `{{ repo_url }}` is the current URL for the repository and `{{ revision }}` is the branch on which the pipeline run was started. See the "Parameters in a pipeline run specification" section for a full list of the parameters.

Reviewer 2: ```suggestion
. Optional: Use parameters in your pipeline run specification. The {pac} resolver replaces these parameters with values that represent the current context. For example, `{{ repo_url }}` is the current URL for the repository and `{{ revision }}` is the commit SHA on which the pipeline run was started. See the "Parameters in a pipeline run specification" section for a full list of the parameters.
```

revision is commit SHA not a branch

> . Optional: Use parameters in your pipeline run specification. The {pac} resolver replaces these parameters with values that represent the current context. For example, `{{ repo_url }}` is the current URL for the repository and `{{ revision }}` is the branch on which the pipeline run was started. See the "Parameters in a pipeline run specification" section for a full list of the parameters.

Author: Will do, thanks

> . Optional: Use parameters in your pipeline run specification. The {pac} resolver replaces these parameters with values that represent the current context. For example, `{{ repo_url }}` is the current URL for the repository and `{{ revision }}` is the branch on which the pipeline run was started. See the "Parameters in a pipeline run specification" section for a full list of the parameters.

Author: I thought the OCP guidelines called for .titles for examples but I might be wrong here

+
.Example of using parameters
[source,yaml]
----
apiVersion: tekton.dev/v1
kind: PipelineRun
metadata:
  name: build-plr
spec:
  pipelineSpec:
    tasks:
      - name: fetch-repo
        taskRef:
          - name: git-clone
        params:
        - name: url
          value: {{ repo_url }}
        - name: revision
          value: {{ revision }}
        workspaces:
        - name: output
          workspace: shared-workspace
# ...
----

> . Optional: Create other files with a `.yaml` or `.yml` extension in the `.tekton` directory and provide definitions of resources that your pipeline run definition references, such as `Pipeline`, `Task`, or `StepAction` CRs. The {pac} resolver automatically includes these resources when creating a `PipelineRun` CR based on your definition.

Reviewer 2: ```suggestion
. Optional: Create other files with a `.yaml` or `.yml` extension in the `.tekton` directory and provide definitions of resources that your pipeline run definition references, such as `Pipeline`, `Task`, or `StepAction` CRs. The {pac} resolver automatically resolves these resources when creating a `PipelineRun` CR based on your definition.
```

Author: "Resolves" on its own does not seem to convey the idea that the actual CR includes the resource definition. I tried a compromise, hope it works.

> . Optional: Add one or more {pac} resolver annotations to the pipeline run definition file. A {pac} resolver annotation references a pipeline or task in {tekton-hub}, an HTTP location, or a location in your repository outside the `.tekton` directory. If you create a {pac} resolver annotation to reference a resource, you can use this resource by name in your pipeline run definition. The {pac} resolver automatically includes the referenced resources when creating a `PipelineRun` CR based on your definition.

Reviewer 2: ```suggestion
. Optional: Add one or more {pac} resolver annotations to the pipeline run definition file. A {pac} resolver annotation references a pipeline or task in {tekton-hub}, an HTTP location, or a location in your repository outside the `.tekton` directory. If you create a {pac} resolver annotation to reference a resource, you can use this resource by name in your pipeline run definition. The {pac} resolver automatically resolves the referenced resources when creating a `PipelineRun` CR based on your definition.
```

Author: same here

> . Optional: Add one or more {pac} resolver annotations to the pipeline run definition file. A {pac} resolver annotation references a pipeline or task in {tekton-hub}, an HTTP location, or a location in your repository outside the `.tekton` directory. If you create a {pac} resolver annotation to reference a resource, you can use this resource by name in your pipeline run definition. The {pac} resolver automatically includes the referenced resources when creating a `PipelineRun` CR based on your definition.

Author: good point. I already put brief cocneptuat information at the very start of the PAC chapter. but some repetitiveness right here won't actually hurt and putting it at the start might work better, yes (especially since in the upcoming DITA that part is outright called context)

+

> See the "{pac} resolver annotations" section for detailed information about these annotations.

Author: Unfortunately the xref part is not actually allowed.

+
.Example of using a {pac} resolver annotation to reference a task in {tekton-hub}
[source,yaml]
----
apiVersion: tekton.dev/v1
kind: PipelineRun
metadata:
  name: build-plr
annotations:

>   pipelinesascode.tekton.dev/task: "[git-clone]"

Author: I certainly am open to callouts. We can't proivide YAML-incomplete exampels beacuse of OCP guidelines, so callouts can help highlight what is actually important, I guess.

spec:
  pipelineSpec:
    workspaces:
    - name: shared-workspace
      emptyDir: {}
    tasks:
      - name: fetch-repo
        taskRef:
          - name: git-clone
        params:
        - name: url
          value: {{ repo_url }}
        - name: revision
          value: {{ revision }}
        workspaces:
        - name: output
          workspace: shared-workspace
# ...
----

> . Define the event that triggers this pipeline run by adding one of the following sets of annotations. See the "Annotations for matching events to a pipeline run" section for detailed information about these annotations.

Author: Unfortunately, it's "sets", because "on-event" and "on-target-branch" must come together. It would be better to have a clearer term, though. Maybe even "any of the following" might work if I still make it clear that these two come together?

> ** A combination of the `pipelinesascode.tekton.dev/on-event` annotation, which defines a `pull request` or `push` event, and a `pipelinesascode.tekton.dev/on-target-branch` annotation, which defines the branch that the pull request or push event must target. If you match your pipeline run to a pull request or push event, the pipeline run also starts each time the source branch of the pull request or push event is updated.

Reviewer 2: ```suggestion
** A combination of the `pipelinesascode.tekton.dev/on-event` annotation, which defines a `pull request` or `push` event, and a `pipelinesascode.tekton.dev/on-target-branch` annotation, which defines the branch that the pull request or push event must target. If your pipeline run is matched to a pull request or push event, the pipeline run also starts each time the source branch of the pull request or push event is updated.
```

Author: The style guides for docs prefer active voice when at all possible.

> ** A combination of the `pipelinesascode.tekton.dev/on-event` annotation, which defines a `pull request` or `push` event, and a `pipelinesascode.tekton.dev/on-target-branch` annotation, which defines the branch that the pull request or push event must target. If you match your pipeline run to a pull request or push event, the pipeline run also starts each time the source branch of the pull request or push event is updated.

Author: Yes, this is one of the three options of what we can add, which is why it's not a complete sentence.

> ** A combination of the `pipelinesascode.tekton.dev/on-event` annotation, which defines a `pull request` or `push` event, and a `pipelinesascode.tekton.dev/on-target-branch` annotation, which defines the branch that the pull request or push event must target. If you match your pipeline run to a pull request or push event, the pipeline run also starts each time the source branch of the pull request or push event is updated.

Author: In addition to when the pull request or push event is first created, which I need to make clear here.

> ** A combination of the `pipelinesascode.tekton.dev/on-event` annotation, which defines a `pull request` or `push` event, and a `pipelinesascode.tekton.dev/on-target-branch` annotation, which defines the branch that the pull request or push event must target. If you match your pipeline run to a pull request or push event, the pipeline run also starts each time the source branch of the pull request or push event is updated.

Reviewer: I thought so but I had to read it so many times, lol And I was also thrown off because the lead in sentence didn't immediately introduce the bulleted list (which [this suggestion](https://github.com/openshift/openshift-docs/pull/89926/files#r1989974739) changes).

+
[NOTE]
====

> If your Git repository provider uses merge requests and not pull request, a `pull_request` event definition matches a merge request.

Reviewer 2: ```suggestion
If you're using GitLab provider, it does use merge request term, a `pull_request` event matches a merge request.
```

@mramendi only GitLab uses term "Merge Request" so I think you fully rephrase this sentence

Reviewer 2: or you can enhance it better I just gave a suggestion.

====
+
.Example of matching a pipeline run to a pull request event
[source,yaml]
----
apiVersion: tekton.dev/v1
kind: PipelineRun
metadata:
  name: pipeline-pr-main
annotations:
  pipelinesascode.tekton.dev/on-target-branch: "[main]"
  pipelinesascode.tekton.dev/on-event: "[pull_request]"
# ...
----

> 

Author: It would not be a lot of work. The examples already are in the reference section, so we're just editing this one module (plus the other concepts module), which is exactly the planned scope

** A `pipelinesascode.tekton.dev/on-comment` annotation, which matches the pipeline run to a comment on a pull request by regular expression. If you match your pipeline run to a comment, it starts when the comment is added to a pull request. To start the pipeline run again. add the comment again.
+
.Example of matching a pipeline run to a comment
[source,yaml]
----
apiVersion: tekton.dev/v1
kind: PipelineRun
metadata:
  name: pipeline-comment
annotations:
  pipelinesascode.tekton.dev/on-comment: "^/merge-pr"
# ...
----

> ** A `pipelinesascode.tekton.dev/on-cel-expression` annotation, which is matched if the pull request or push event matches the specified Common Expression Language (CEL) expression.

Reviewer 2: ```suggestion
** A `pipelinesascode.tekton.dev/on-cel-expression` annotation, matches pipeline run if Common Expression Language (CEL) expression is evaluated true on a pull request or push event.
```

Author: did something :)

. Optional: Add one or more annotations that filter the matching events. With these annotations, when the defined matching event (such as a pull request, a push event. or a comment) happens, {pac} checks if these annotations are also matched. The pipeline run starts only if all the annotations that you added to it are matched. See the "Annotations for filtering events" section for detailed information about these annotations.

** A `pipelinesascode.tekton.dev/on-path-changed` annotation is matched if the pull request or push event affect files in the specified paths.

> ** A `pipelinesascode.tekton.dev/on-path-changed-ignore` annotation excludes matching the event if it affects _only_ files in the specified paths.

Reviewer 2: @mramendi I think it should be rephrased, at the moment it seems vague to me.

> ** A `pipelinesascode.tekton.dev/on-cel-expression` annotation is matched if the pull request or push event matches a Common Expression Language (CEL) expression.

Reviewer 3: Can we add a note saying that if annothation is provided and same set of key value provided as part of on-cel-expression then which one take precedence :thinking:

Reviewer 2: @mramendi `on-cel-expression` will take precedence on `on-event, on-target-branch, on-label, on-path-change, and on-path-change-ignore` annotations.

Author: Added to the reference part - and I assumed this leads to completely ignoring the other annotations, I might be wrong, please review there

Reviewer 2: @mramendi  yes, if `on-cel-expression` annotation is defined in the PipelineRun definition then others will be ignored.

Reviewer 4: it's all documented in the upstream documentation already

![image](https://github.com/user-attachments/assets/0ad7dd4e-1f51-48dc-adbe-226a6da46ef9)

+
.Example of using annotations to filter matching events in a pipeline run
[source,yaml]
----
apiVersion: tekton.dev/v1
kind: PipelineRun
metadata:
  name: pipeline-pr-filtered
annotations:
  pipelinesascode.tekton.dev/on-target-branch: "[main]"
  pipelinesascode.tekton.dev/on-event: "[pull_request]"
  pipelinesascode.tekton.dev/on-label: "[bug, defect]"
  pipelinesascode.tekton.dev/on-path-changed: "["src/*", "cli/**"]"
# ...
----

> . Optional: Add the `pipelinesascode.tekton.dev/cancel-in-progress: "true"` annotation to enable automatic cancellation of the pipeline run in certain cases. For example, if a pull request triggers a pipeline run and then the user pushes new commits into the pull request source branch, each push triggers a new copy of the pipeline run. if you enable automatic cancellation-in-progress, after a new copy of the pipeline run starts, {pac} cancels the older run ti avoid running many copies of the pipeline run at the same time. See the "Annotations for specifying automatic cancellation-in-progress for a pipeline run" for detailed information about this annotation.

Reviewer 3: ```suggestion
. Optional: Add the `pipelinesascode.tekton.dev/cancel-in-progress: "true"` annotation to enable automatic cancellation of the pipeline run in certain cases. For example, if a pull request triggers a pipeline run and then the user pushes new commits into the pull request source branch, each push triggers a new copy of the pipeline run. if you enable automatic cancellation-in-progress, after a new copy of the pipeline run starts, {pac} cancels the older run to avoid running many copies of the pipeline run at the same time. See the "Annotations for specifying automatic cancellation-in-progress for a pipeline run" for detailed information about this annotation.
```
